File: android/src/main/java/com/margelo/nitro/rive/DataSourceResolver.kt
package com.margelo.nitro.rive

import android.net.Uri
import java.net.URI

object DataSourceResolver {
  fun resolve(asset: ResolvedReferencedAsset): DataSource? {
    asset.sourceUrl?.let { return resolveFromUrl(it) }
    asset.sourceAssetId?.let { return resolveFromAssetId(it) }
    asset.sourceAsset?.let { return DataSource.resource(it) }
    return null
  }

  private fun resolveFromUrl(urlString: String): DataSource {
    val uri = runCatching { URI(urlString) }.getOrElse {
      throw DataLoaderException.InvalidURL(urlString)
    }

    return when (uri.scheme) {
      "file" -> uri.path?.let { DataSource.File(it) }
        ?: throw DataLoaderException.InvalidURL(urlString)
      "http", "https" -> DataSource.Http(urlString)
      else -> throw DataLoaderException.InvalidURL(urlString)
    }
  }

  private fun resolveFromAssetId(assetId: String): DataSource {
    val scheme = runCatching { Uri.parse(assetId).scheme }.getOrNull()

    if (scheme != null) {
      return resolveFromUrl(assetId)
    }

    return DataSource.Resource(assetId)
  }
}

File: ios/DataSourceResolver.swift
import Foundation

struct DataSourceResolver {
  static func resolve(from asset: ResolvedReferencedAsset) throws -> DataSource? {
    if let sourceUrl = asset.sourceUrl {
      return try resolveFromUrl(sourceUrl)
    }

    if let sourceAssetId = asset.sourceAssetId {
      return try resolveFromAssetId(sourceAssetId)
    }

    if let sourceAsset = asset.sourceAsset {
      return resolveFromSourceAsset(sourceAsset)
    }

    return nil
  }

  private static func resolveFromUrl(_ urlString: String) throws -> DataSource {
    guard let url = URL(string: urlString) else {
      throw DataLoaderError.invalidURL(urlString)
    }

    if url.isFileURL {
      return .file(url: url)
    }

    if url.scheme == "http" || url.scheme == "https" {
      return .http(url: url)
    }

    throw DataLoaderError.invalidURL(urlString)
  }

  private static func resolveFromAssetId(_ assetId: String) throws -> DataSource {
    guard let url = URL(string: assetId) else {
      throw DataLoaderError.invalidURL(assetId)
    }

    if url.isFileURL {
      return .file(url: url)
    }

    if url.scheme == "http" || url.scheme == "https" {
      return .http(url: url)
    }

    throw DataLoaderError.invalidURL(assetId)
  }

  private static func resolveFromSourceAsset(_ sourceAsset: String) -> DataSource {
    return .bundle(nameWithExtension: sourceAsset)
  }
}
